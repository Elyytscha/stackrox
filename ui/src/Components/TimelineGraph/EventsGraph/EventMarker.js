import React from 'react';
import { scaleLinear } from 'd3-scale';

import selectors from 'Components/TimelineGraph/MainView/selectors';
import { getWidth } from 'utils/d3Utils';
import D3Anchor from 'Components/D3Anchor';

const EventMarker = ({
    differenceInHours,
    translateX,
    translateY,
    minTimeRange,
    maxTimeRange,
    size
}) => {
    // the "container" argument is a reference to the container for the D3-related element
    function onUpdate(container) {
        const width = getWidth(selectors.svgSelector);
        const xScale = scaleLinear()
            .domain([minTimeRange, maxTimeRange])
            .range([0, width]);
        const x = xScale(differenceInHours).toFixed(0);

        container.attr(
            'transform',
            `translate(${Number(translateX) + Number(x) - size / 2}, ${Number(translateY) -
                size / 2})`
        );
    }
    return (
        <D3Anchor
            dataTestId="timeline-event-marker"
            translateX={translateX}
            translateY={translateY}
            onUpdate={onUpdate}
        >
            <rect fill="var(--primary-600)" width={size} height={size} />
        </D3Anchor>
    );
};

export default EventMarker;
