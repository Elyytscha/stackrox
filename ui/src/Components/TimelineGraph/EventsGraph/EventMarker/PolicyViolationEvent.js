import React from 'react';
import PropTypes from 'prop-types';

const PolicyViolationEvent = ({ width, height }) => {
    const elementHeight = height || width;
    return (
        <rect
            data-testid="policy-violation-event"
            fill="var(--primary-600)"
            width={width}
            height={elementHeight}
            rx={3}
        />
    );
};

PolicyViolationEvent.propTypes = {
    width: PropTypes.number.isRequired,
    height: PropTypes.number
};

PolicyViolationEvent.defaultProps = {
    height: null
};

export default PolicyViolationEvent;
