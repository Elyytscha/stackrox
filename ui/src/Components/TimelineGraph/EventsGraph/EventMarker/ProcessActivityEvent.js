import React from 'react';
import PropTypes from 'prop-types';

const ProcessActivityEvent = ({ width, height }) => {
    const elementHeight = height || width;
    return (
        <polygon
            data-testid="process-activity-event"
            points={`${width / 2},0 0,${elementHeight / 2} ${width /
                2},${elementHeight} ${width},${elementHeight / 2}`}
            fill="var(--alert-600)"
            stroke="var(--alert-600)"
        />
    );
};

ProcessActivityEvent.propTypes = {
    width: PropTypes.number.isRequired,
    height: PropTypes.number
};

ProcessActivityEvent.defaultProps = {
    height: null
};

export default ProcessActivityEvent;
