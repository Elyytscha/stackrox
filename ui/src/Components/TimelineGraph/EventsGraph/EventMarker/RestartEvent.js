import React from 'react';
import PropTypes from 'prop-types';

const RestartEvent = ({ height, width }) => {
    const elementHeight = height || width;
    return (
        <polygon
            data-testid="restart-event"
            points={`0,${elementHeight} ${width / 2},0 ${width},${elementHeight}`}
            fill="var(--caution-600)"
            stroke="var(--caution-600)"
        />
    );
};

RestartEvent.propTypes = {
    width: PropTypes.number.isRequired,
    height: PropTypes.number
};

RestartEvent.defaultProps = {
    height: null
};

export default RestartEvent;
