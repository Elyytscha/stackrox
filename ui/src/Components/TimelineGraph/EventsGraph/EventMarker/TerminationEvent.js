import React from 'react';
import PropTypes from 'prop-types';

const TerminationEvent = ({ height, width }) => {
    const elementHeight = height || width;
    return (
        <polygon
            data-testid="termination-event"
            points={`0,0 ${width / 2},${elementHeight} ${width},0`}
            fill="var(--caution-600)"
            stroke="var(--caution-600)"
        />
    );
};

TerminationEvent.propTypes = {
    width: PropTypes.number.isRequired,
    height: PropTypes.number
};

TerminationEvent.defaultProps = {
    height: null
};

export default TerminationEvent;
