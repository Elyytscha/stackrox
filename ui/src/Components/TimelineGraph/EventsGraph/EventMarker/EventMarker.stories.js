/* eslint-disable react-hooks/rules-of-hooks */
import React from 'react';

import EventMarker from './EventMarker';

export default {
    title: 'EventMarker',
    component: EventMarker,
};

export const policyViolationEvent = () => {
    return (
        <svg data-testid="timeline-main-view" width={50} height={50}>
            <EventMarker
                name="event"
                type="PolicyViolationEvent"
                timestamp="2020-04-20T20:20:20.358227916Z"
                differenceInMilliseconds={50}
                translateX={0}
                translateY={25}
                minTimeRange={0}
                maxTimeRange={100}
                size={20}
            />
        </svg>
    );
};

export const processActivityEvent = () => {
    return (
        <svg data-testid="timeline-main-view" width={50} height={50}>
            <EventMarker
                name="event"
                type="ProcessActivityEvent"
                uid={0}
                timestamp="2020-04-20T20:20:20.358227916Z"
                differenceInMilliseconds={50}
                translateX={0}
                translateY={25}
                minTimeRange={0}
                maxTimeRange={100}
                size={20}
            />
        </svg>
    );
};

export const whitelistedProcessActivityEvent = () => {
    return (
        <svg data-testid="timeline-main-view" width={50} height={50}>
            <EventMarker
                name="event"
                type="ProcessActivityEvent"
                uid={0}
                whitelisted
                timestamp="2020-04-20T20:20:20.358227916Z"
                differenceInMilliseconds={50}
                translateX={0}
                translateY={25}
                minTimeRange={0}
                maxTimeRange={100}
                size={20}
            />
        </svg>
    );
};

export const restartEvent = () => {
    return (
        <svg data-testid="timeline-main-view" width={50} height={50}>
            <EventMarker
                name="event"
                type="ContainerRestartEvent"
                timestamp="2020-04-20T20:20:20.358227916Z"
                differenceInMilliseconds={50}
                translateX={0}
                translateY={25}
                minTimeRange={0}
                maxTimeRange={100}
                size={20}
            />
        </svg>
    );
};

export const terminationEvent = () => {
    return (
        <svg data-testid="timeline-main-view" width={50} height={50}>
            <EventMarker
                name="event"
                type="ContainerTerminationEvent"
                reason="Because of Covid-19"
                timestamp="2020-04-20T20:20:20.358227916Z"
                differenceInMilliseconds={50}
                translateX={0}
                translateY={25}
                minTimeRange={0}
                maxTimeRange={100}
                size={20}
            />
        </svg>
    );
};
