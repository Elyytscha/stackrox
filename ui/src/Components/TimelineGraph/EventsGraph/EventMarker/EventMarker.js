import React from 'react';
import { scaleLinear } from 'd3-scale';

import { getWidth } from 'utils/d3Utils';
import { eventTypes } from 'constants/timelineTypes';
import selectors from 'Components/TimelineGraph/MainView/selectors';
import D3Anchor from 'Components/D3Anchor';
import PolicyViolationEvent from './PolicyViolationEvent';
import RestartEvent from './RestartEvent';
import ProcessActivityEvent from './ProcessActivityEvent';
import TerminationEvent from './TerminationEvent';

const EventMarker = ({
    type,
    differenceInHours,
    translateX,
    translateY,
    minTimeRange,
    maxTimeRange,
    size
}) => {
    // the "container" argument is a reference to the container for the D3-related element
    function onUpdate(container) {
        const width = getWidth(selectors.svgSelector);
        const xScale = scaleLinear()
            .domain([minTimeRange, maxTimeRange])
            .range([0, width]);
        const x = xScale(differenceInHours).toFixed(0);

        container.attr(
            'transform',
            `translate(${Number(translateX) + Number(x) - size / 2}, ${Number(translateY) -
                size / 2})`
        );
    }

    return (
        <D3Anchor
            dataTestId="timeline-event-marker"
            translateX={translateX}
            translateY={translateY}
            onUpdate={onUpdate}
        >
            {type === eventTypes.POLICY_VIOLATION && <PolicyViolationEvent width={size} />}
            {type === eventTypes.PROCESS_ACTIVITY && <ProcessActivityEvent width={size} />}
            {type === eventTypes.RESTART && <RestartEvent width={size} />}
            {type === eventTypes.TERMINATION && <TerminationEvent width={size} />}
        </D3Anchor>
    );
};

export default EventMarker;
