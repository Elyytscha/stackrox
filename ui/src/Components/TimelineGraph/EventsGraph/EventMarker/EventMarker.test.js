import React from 'react';
import { render } from '@testing-library/react';

import EventMarker from './EventMarker';

test('should show a policy violation event marker', async () => {
    const { queryByTestId, asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                name="eventName"
                type="PolicyViolationEvent"
                differenceInHours={1}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={2}
            />
        </svg>
    );
    expect(queryByTestId('policy-violation-event')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
});

test('should show a process activity event marker', async () => {
    const { queryByTestId, asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                name="eventName"
                type="ProcessActivityEvent"
                differenceInHours={1}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={2}
            />
        </svg>
    );
    expect(queryByTestId('process-activity-event')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
});

test('should show a container restart event marker', async () => {
    const { queryByTestId, asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                name="eventName"
                type="ContainerRestartEvent"
                differenceInHours={1}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={2}
            />
        </svg>
    );
    expect(queryByTestId('restart-event')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
});

test('should show a container termination event marker', async () => {
    const { queryByTestId, asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                name="eventName"
                type="ContainerTerminationEvent"
                differenceInHours={1}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={2}
            />
        </svg>
    );
    expect(queryByTestId('termination-event')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
});
