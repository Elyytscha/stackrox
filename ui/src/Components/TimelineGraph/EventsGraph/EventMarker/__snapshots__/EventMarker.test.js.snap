// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a container restart event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <polygon
        data-testid="restart-event"
        fill="var(--caution-600)"
        points="0,10 5,0 10,10"
        stroke="var(--caution-600)"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`should show a container termination event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <polygon
        data-testid="termination-event"
        fill="var(--caution-600)"
        points="0,0 5,10 10,0"
        stroke="var(--caution-600)"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`should show a policy violation event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <rect
        data-testid="policy-violation-event"
        fill="var(--primary-600)"
        height="10"
        rx="3"
        width="10"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`should show a process activity event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <polygon
        data-testid="process-activity-event"
        fill="var(--alert-600)"
        points="5,0 0,5 5,10 10,5"
        stroke="var(--alert-600)"
      />
    </g>
  </svg>
</DocumentFragment>
`;
